File: Java/NQueens.java
class Solution {

    private Set<Integer> cols = new HashSet<>();
    private Set<Integer> diff = new HashSet<>();
    private Set<Integer> sum = new HashSet<>();

    public List<List<String>> solveNQueens(int n) {
        List<List<String>> res = new ArrayList<>();
        if(n <= 0) return res;

        char[][] board = new char[n][n];
        initBoard(board, n);

        dfs(n, 0, board, res);

        return res;
    }

    private void dfs(int n, int row, char[][]board, List<List<String>> res) {
        if(row == n) {
            List<String> tmp = new ArrayList<>();
            for(int i = 0; i < n; i++) {
                tmp.add(String.valueOf(board[i]));
            }
            res.add(tmp);
            return;
        }

        for(int col = 0; col < n; col++) {
            if(!cols.contains(col) && !diff.contains(col - row) && !sum.contains(col + row)) {
                cols.add(col); diff.add(col - row); sum.add(col + row);
                board[row][col] = 'Q';
                dfs(n, row + 1, board, res);
                cols.remove(col); diff.remove(col - row); sum.remove(col + row);
                board[row][col] = '.';
            }
        }

    }

    private void initBoard(char[][] board, int n) {
        for(int i = 0; i < n; i++) {
            for(int j = 0; j < n; j++) {
                board[i][j] = '.';
            }
        }
    }
}

File: Go/nqueens.go
func solveNQueens(n int) [][]string {
    var res [][]string
    if  n <= 0 {return res}

    cols := make(map[int]int, n)
    diff := make(map[int]int, n)
    sum := make(map[int]int, n)
    board := make([][]rune, n)
	for i := 0; i < n; i++ {
		item := make([]rune, n)
		for j := 0; j < n; j++ {
			item[j] = '.'
		}
		board[i] = item
	}

    var dfs func(row int)
    dfs = func(row int) {
        if row == n {
            var tmp []string
            for _, v := range board {
                tmp = append(tmp, string(v))
            }
            res = append(res, tmp)
            return;
        }
        for col := 0; col < n; col++ {
            if cols[col] == 0 && diff[col - row] == 0 && sum[col + row] == 0 {
                cols[col] = 1
                diff[col - row] = 1
                sum[col + row] = 1
                board[row][col] = 'Q'
                dfs(row + 1)
                cols[col] = 0
                diff[col - row] = 0
                sum[col + row] = 0
                board[row][col] = '.'
            }
        }
    }
    
    dfs(0)
    return res
}